lightning/src/onion_message/messenger.rs

```diff
@@ -1436,12 +1436,30 @@ where
 				match message {
 					ParsedOnionMessageContents::Offers(msg) => {
 						let responder = reply_path.map(Responder::new);
-						let response_instructions = self.offers_handler.handle_message(msg, responder);
+						let context = match context {
+							None => OffersContext::Unknown {},
+							Some(MessageContext::Offers(context)) => context,
+							Some(MessageContext::Custom(_)) => {
+								debug_assert!(false, "Shouldn't have triggered this case.");
+								return
+							}
```

Comment:
Could you remind me why we can't do this as part of `Payload` parsing? I think we discussed it, but it would be nice to avoid this case if we can handle it earlier.

Reply:
So the way we have set it up ensures that only the correct (`msg_type`, `recipient_data_type`) combinations are parsed in `peel_onion_message`. The checks we're performing in `handle_onion_message` serve as an additional fail-safe to ensure that this scenario isn’t ever triggered, whether due to a mistake on our end or a malicious attempt by someone else.

One quick question: since we expect this condition to never actually occur, do you think we should implement a stricter response, such as using an explicit `panic!()`?
Let me know what you think!

Thanks!

Reply:
> So the way we have set it up ensures that only the correct (`msg_type`, `recipient_data_type`) combinations are parsed in `peel_onion_message`. The checks we're performing in `handle_onion_message` serve as an additional fail-safe to ensure that this scenario isn’t ever triggered, whether due to a mistake on our end or a malicious attempt by someone else.

Right, but that doesn't answer my question. More precisely, was there a reason why we aren't able to include the context data as another part of `ParsedOnionMessageContents`? (i.e. have each variant wrap both a message and a context instead of having the context returned separately). That would avoid the problem entirely.

Here we are returning a `Payload<ParsedOnionMessageContents<..>>` which spreads the the message and context across two different fields, one being an enum with a variant containing the `ReceiveTlvs`. Maybe that is why?

https://github.com/lightningdevkit/rust-lightning/blob/88e1b56d66ff550b36a6d422f47c9b9729406f61/lightning/src/onion_message/packet.rs#L243

@TheBlueMatt Any opinon on this? Would it be worth refactoring (if not now, then in a follow-up)?

Reply:
Spoke offline. Let's leave this to a follow-up.